runtime/signals.c

```diff
@@ -465,54 +468,67 @@ CAMLexport int caml_rev_convert_signal_number(int signo)
   return signo;
 }
 
-void * caml_init_signal_stack(void)
+stack_t * caml_init_signal_stack(void)
 {
 #ifdef POSIX_SIGNALS
-  stack_t stk;
-  stk.ss_flags = 0;
-  stk.ss_size = SIGSTKSZ;
+  stack_t * stk = malloc(sizeof(stack_t));
+  stk->ss_flags = 0;
+  stk->ss_size = SIGSTKSZ;
   /* The memory used for the alternate signal stack must not free'd before
-     calling sigaltstack with SS_DISABLE. malloc is therefore used rather
+     calling sigaltstack with SS_DISABLE. malloc/mmap is therefore used rather
      than caml_stat_alloc_noexc so that if a shutdown path erroneously fails
      to call caml_free_signal_stack then we have a memory leak rather than a
      nasty piece of undefined behaviour forced on the caller. */
-  stk.ss_sp = malloc(stk.ss_size);
-  if(stk.ss_sp == NULL) {
+#ifdef USE_MMAP_MAP_STACK
+  stk->ss_sp = mmap(NULL, stk->ss_size, PROT_WRITE | PROT_READ,
+             MAP_ANONYMOUS | MAP_PRIVATE | MAP_STACK, -1, 0);
+  if (stk->ss_sp == MAP_FAILED)
     return NULL;
-  }
-  if (sigaltstack(&stk, NULL) < 0) {
-    free(stk.ss_sp);
+#else
+  stk->ss_sp = malloc(stk->ss_size);
+  if(stk->ss_sp == NULL)
+    return NULL;
+#endif /* USE_MMAP_MAP_STACK */
+
+  if (sigaltstack(stk, NULL) < 0) {
+    free(stk->ss_sp);
     return NULL;
   }
-  return stk.ss_sp;
+  return stk;
 #else
   return NULL;
 #endif
 }
 
-void caml_free_signal_stack(void * signal_stack)
+void caml_free_signal_stack(stack_t * signal_stack)
 {
 #ifdef POSIX_SIGNALS
-  stack_t stk, disable;
-  disable.ss_flags = SS_DISABLE;
-  disable.ss_sp = NULL;  /* not required but avoids a valgrind false alarm */
-  disable.ss_size = SIGSTKSZ; /* macOS wants a valid size here */
-  if (sigaltstack(&disable, &stk) < 0) {
-    caml_fatal_error("Failed to reset signal stack (err %d)", errno);
+  stack_t stk;
+  /* Get current stack */
+  if (sigaltstack(NULL, &stk) < 0) {
```

Comment:
I don't think this is strictly related to the OpenBSD fix, but on the surface this approach of "get current stack, and disable only if it's ours" looks better than the current "disable the stack and re-enable it if it turns out not to be ours". What do you think, @xavierleroy?

Reply:
I thought about this alternate approach, but it needs two `sigaltstack` calls in the normal case where the alt stack is ours,  while my hack is only one syscall in the normal case and two in the unlikely case.

Reply:
I thought that was very likely the case, thanks!